src/NUnitFramework/framework/Internal/AsyncEnumerableAdapter.NetFramework.cs

```diff
@@ -0,0 +1,112 @@
+            if (asyncEnumerable is null)
+                return false;
+
+            var getEnumeratorMethod = asyncEnumerable.GetMethod("GetAsyncEnumerator");
+            if (getEnumeratorMethod is null)
+                return false;
+
+            var enumeratorType = getEnumeratorMethod.ReturnType;
+
+            var currentProperty = enumeratorType.GetProperty("Current");
+            if (currentProperty is null)
+                return false;
+
+            var moveNextMethod = enumeratorType.GetMethod("MoveNextAsync");
+            if (moveNextMethod is null)
+                return false;
+
+            var asyncDisposableType = enumeratorType.GetInterface("System.IAsyncDisposable");
+            if (asyncDisposableType is null)
+                return false;
+
+            var asyncDisposeMethod = asyncDisposableType!.GetMethod("DisposeAsync");
+            if (asyncDisposeMethod is null)
+                return false;
+
+            var shape = new AsyncEnumerableShapeInfo(
+                getEnumeratorMethod,
+                currentProperty,
+                moveNextMethod,
+                asyncDisposeMethod);
+
+            result = new AsyncEnumerableWrapper(shape, enumerable);
+            return true;
+        }
+
+        private class AsyncEnumerableWrapper : IEnumerable<object?>
+        {
+            private readonly AsyncEnumerableShapeInfo _shape;
+            private readonly object _asyncEnumerable;
+
+            public AsyncEnumerableWrapper(AsyncEnumerableShapeInfo shape, object asyncEnumerable)
+            {
+                _shape = shape;
+                _asyncEnumerable = asyncEnumerable;
+            }
+
+            public IEnumerator<object?> GetEnumerator()
+                => new AsyncEnumeratorWrapper(_shape, _shape.GetAsyncEnumeratorMethod.Invoke(_asyncEnumerable, new object[] { CancellationToken.None })!);
+
+            IEnumerator IEnumerable.GetEnumerator()
+                => new AsyncEnumeratorWrapper(_shape, _shape.GetAsyncEnumeratorMethod.Invoke(_asyncEnumerable, new object[] { CancellationToken.None })!);
+        }
+
+        private class AsyncEnumeratorWrapper : IEnumerator<object?>
+        {
+            private readonly AsyncEnumerableShapeInfo _shape;
+            private readonly object _asyncEnumerator;
+
+            public AsyncEnumeratorWrapper(AsyncEnumerableShapeInfo shape, object asyncEnumerator)
+            {
+                _shape = shape;
+                _asyncEnumerator = asyncEnumerator;
+            }
+
+            public object? Current => _shape.CurrentProperty.GetValue(_asyncEnumerator);
+
+            public void Dispose()
+                => AsyncToSyncAdapter.Await(() => _shape.DisposeAsyncMethod.Invoke(_asyncEnumerator, null));
+
+            public bool MoveNext()
+                => AsyncToSyncAdapter.Await<bool>(() => _shape.MoveNextAsyncMethod.Invoke(_asyncEnumerator, null));
```

Comment:
MethodInfo.Invoke is sloow. As the signature is known, we could use delegates:
```
Job=.NET Framework 4.6.2  Runtime=.NET Framework 4.6.2

| Method               | Mean       | Ratio | Allocated | Alloc Ratio |
|--------------------- |-----------:|------:|----------:|------------:|
| DirectMethodCall     |   1.145 ns |  0.01 |         - |        0.00 |
| MethodInfoInvoke     | 102.406 ns |  1.00 |     112 B |        1.00 |
| CreateDelegateInvoke |   2.052 ns |  0.02 |         - |        0.00 |
```

Something like:

```csharp
private delegate ValueTask<bool> MoveNextAsyncDelegate(object @this);

...
MoveNextAsync = (MoveNextAsyncDelegate)moveNextAsyncMethod.CreateDelegate(typeof(MoveNextAsyncDelegate));
...

=> AsyncToSyncAdapter.Await<bool>(() => _shape.MoveNextAsync(_asyncEnumerator));
```

Although all this is probably a mute point as long as we need to use the `AsyncToSyncAdapter.Await`

Reply:
I like both of these ideas, thanks, though I agree top-level async may need to wait 😄 

I think I'd mistaken the overhead of `CreateDelegate()` with the overhead of using `Reflection.Emit`-based codegen but I can see that it's comparatively comparatively quite cheap even on [netfx](https://referencesource.microsoft.com/#mscorlib/system/delegate.cs,715).

Reply:
I have a few ideas here that'll take a bit to experiment with. I think changes for the rest of the suggestions have been pushed